### PR TITLE
Add WorkspaceDataFactory with traits for Tuples and WS

### PR DESCRIPTION
**Pull request by klecki**
Some common manipulation methods for all
of the output types of Ops

Next #540

### CODE REVIEW

File: dali/pipeline/executor/executor.cc [hunk 1]
```diff
@@ -467,8 +467,8 @@ void Executor::SetupDataForGraph(WorkspaceBlob *wsb) {
   }
 
   // Setup mixed op input and output buffers
-  for (int i = 0; i < graph_->NumMixedOp(); ++i) {
-    OpNode &node = graph_->mixed_node(i);
+  for (int i = 0; i < graph_->NumOp(OpType::MIXED); ++i) {
+    OpNode &node = graph_->Node(OpType::MIXED, i);
     MixedWorkspace &ws = wsb->mixed_op_data[i];
 
     // Mixed ops do not take argument tensors (at least for now)
```
Comment (jantonguirao) [inline dali/pipeline/executor/executor.cc]: Maybe unrelated to this PR but inportant for new nvJPEG fused crop operators:
Can we make mixed operators accept argument tensors as well?

File: dali/pipeline/util/op_type_to_workspace.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  using type = MixedWorkspace;
+};
+
+template <>
+struct workspace_type<OpType::SUPPORT> {
+  using type = SupportWorkspace;
+};
+
+template <OpType op_type>
+using workspace_t = typename workspace_type<op_type>::type;
```
Comment (mzient) [inline dali/pipeline/util/op_type_to_workspace.h]: `workspace_t` suggest something more fundamental than a mapping from a particular enum value to a workspace type. How about `WorkspaceForOpType` or `OpTypeToWorkspace`?

Comment (klecki) [reply dali/pipeline/util/op_type_to_workspace.h]: Done, but still with snake case, as it is a type trait.

**Comment by klecki** [comment]
@mzient  Done

**Comment by klecki** [comment]
Build 665178

**Comment by klecki** [comment]
Build 665963

File: dali/pipeline/workspace/workspace_data_factory.h [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+bool IsPinned(DeviceWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
```
Comment (jantonguirao) [inline dali/pipeline/workspace/workspace_data_factory.h]: why do you need the batch_size?

File: dali/pipeline/workspace/workspace_data_factory.h [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+using tensor_store_elem_t = workspace_out_data_t<GetTensorStoreIndex(op_type, device)>;
+
+// using tensor_data_store_t =
+//     std::tuple<workspace_out_data_t<0>, workspace_out_data_t<1>, workspace_out_data_t<2>,
+//                workspace_out_data_t<3>, workspace_out_data_t<4>, workspace_out_data_t<5>,
+//                workspace_out_data_t<6>, workspace_out_data_t<7>>;
+
+// Tuple containing Workspace Output Type generated for all values
+using tensor_data_store_t = detail::tuple_generator_t<workspace_out_data_t,
+                                                  detail::build_seq_t<0, GetMaxTensorStoreIndex()>>;
```
Comment (jantonguirao) [inline dali/pipeline/workspace/workspace_data_factory.h]: indent

File: dali/pipeline/workspace/workspace_data_factory.h [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+};
+
+template <int storage_idx>
+using workspace_out_data_t = typename workspace_out_data_type_gen<storage_idx>::type;
+
+// OpType -> StorageDevice -> Workspace Output Type
+template <OpType op_type, StorageDevice device>
+using tensor_store_elem_t = workspace_out_data_t<GetTensorStoreIndex(op_type, device)>;
+
+// using tensor_data_store_t =
```
Comment (JanuszL) [inline dali/pipeline/workspace/workspace_data_factory.h]: ?

File: dali/pipeline/workspace/workspace_data_factory.h [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+};
+
+template <StorageDevice device>
+struct BatchFactoryImpl<OpType::SUPPORT, device> {
+  static tensor_store_elem_t<OpType::SUPPORT, device> CreateOutputBatch(
+      int batch_size) {
+    DALI_ENFORCE(device == StorageDevice::CPU, "Only CPU outputs allowed");
+    tensor_store_elem_t<OpType::SUPPORT, device> output(
+        new Tensor<storage_backend_t<device>>);
+    // output->set_pinned(false);
```
Comment (JanuszL) [inline dali/pipeline/workspace/workspace_data_factory.h]: ?

File: dali/pipeline/workspace/workspace_data_factory.h [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+struct BatchFactoryImpl<OpType::CPU, device> {
+  static tensor_store_elem_t<OpType::CPU, device> CreateOutputBatch(
+      int batch_size) {
+    DALI_ENFORCE(device == StorageDevice::CPU, "Only CPU outputs allowed");
+    // Allocate `batch_size` Tensors for this ops
+    // results and add them to the workspace.
+    tensor_store_elem_t<OpType::CPU, device> output(batch_size, nullptr);
+    for (auto &tensor_ptr : output) {
+      tensor_ptr.reset(new Tensor<storage_backend_t<device>>);
+      // tensor_ptr->set_pinned(false);
```
Comment (JanuszL) [inline dali/pipeline/workspace/workspace_data_factory.h]: ?

File: dali/pipeline/workspace/workspace_data_factory.h (hunk 3 above)
Comment (klecki) [reply dali/pipeline/workspace/workspace_data_factory.h]: The batch size is needed when we work with CPU Op outputs. The are vector of `batch_size` Tensors.

File: dali/pipeline/workspace/workspace_data_factory.h (hunk 5 above)
Comment (klecki) [reply dali/pipeline/workspace/workspace_data_factory.h]: Done.

File: dali/pipeline/workspace/workspace_data_factory.h (hunk 4 above)
Comment (klecki) [reply dali/pipeline/workspace/workspace_data_factory.h]: Done.

File: dali/pipeline/workspace/workspace_data_factory.h (hunk 7 above)
Comment (klecki) [reply dali/pipeline/workspace/workspace_data_factory.h]: Left from before I had SetPinned() functions.

File: dali/pipeline/workspace/workspace_data_factory.h (hunk 6 above)
Comment (klecki) [reply dali/pipeline/workspace/workspace_data_factory.h]: Done.

**Pull request merged**